src/background/api.ts

```diff
@@ -1,93 +1,78 @@
+  | { ok: false; reason: 'network down'; details: string }
 
+export async function fetchRoar<T extends object>(path: string, init: RequestInit, decoder: JsonDecoder.Decoder<T>): Promise<FetchRoarResult<T>> {
   try {
-    dispatchBackgroundActions.postTweetStart({ targetId })
-    const res = await makeTweetRequest(tweetFormData(target))
-    if (res.status !== 201) {
-      return dispatchBackgroundActions.postTweetFailure({
-        targetId,
-        error: { message: await res.text() },
-      })
+    const response = await fetch(`${window.roarServerUrl}/${path}`, { credentials: 'include', ...init })
+
+    if (response.status >= 200 && response.status < 300) {
+      const text = await response.text()
+      try {
+        const json = JSON.parse(text)
+        const decoded = decoder.decode(json)
+        if (decoded.isOk()) {
+          return { ok: true, data: decoded.value }
+        } else {
+          return { ok: false, reason: 'response not expected data type', text, details: decoded.error }
+        }
+      } catch (error) {
+        return { ok: false, reason: 'response not json', details: error.message, text }
+      }
+    }
+    if (response.status === 400) {
+      const details = await response.text()
+      return { ok: false, reason: 'bad request', details }
     }
-    const tweetResult = await res.json()
-    if (!tweetResult.url) {
-      return dispatchBackgroundActions.postTweetFailure({
-        targetId,
-        error: { message: 'Response must include a url' },
-      })
+    if (response.status === 401) {
+      return { ok: false, reason: 'unauthorized', details: '401 received' }
     }
-    chrome.tabs.create({ url: tweetResult.url, active: true })
-    return dispatchBackgroundActions.postTweetSuccess({ targetId })
+    if (response.status >= 500) {
+      const details = await response.text()
+      return { ok: false, reason: 'server down', details }
+    }
+    const details = await response.text()
+    return { ok: false, reason: 'unknown status', details, status: response.status }
   } catch (error) {
-    return dispatchBackgroundActions.postTweetFailure({ targetId, error })
+    return { ok: false, reason: 'network down', details: error.message }
   }
 }
 
-function makeHandleRequest(domain: string): Promise<Response> {
-  const params = { domain }
-  const requestURL = new URL('v1/website', window.roarServerUrl)
-  requestURL.search = new URLSearchParams(params).toString()
-  return fetch(requestURL.toString())
+type FeedbackResponseData = {
+  url: string
 }
 
-export const fetchTwitterHandle = async (
-  tabId: number,
-  domain: string,
-  dispatchBackgroundActions: Dispatchers<BackgroundAction>,
-  handleCache: TwitterHandleCache
-) => {
-  try {
-    dispatchBackgroundActions.fetchHandleStart({ tabId })
+const FeedbackResponseDataDecoder = JsonDecoder.object<FeedbackResponseData>({ url: JsonDecoder.string }, 'FeedbackResponseData')
```

Comment:
If I understand correctly, we validate the response from the server. I'm not sure I understand why we are doing this. We control the server and we know what the response is so what's the point? What we need to check is the request from the server-side because everyone can send us a request.

Reply:
You're right that this should have no effect in production, but if we change API responses we could run into these decoder errors in production where we'll get useful error messages, which is the point of implementing the decoders.